algorithms/src/snark/marlin/ahp/prover/round_functions/fourth.rs

```diff
@@ -12,58 +12,268 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use core::convert::TryInto;
 use std::collections::BTreeMap;
 
 use crate::{
-    fft::DensePolynomial,
+    fft::{
+        domain::{FFTPrecomputation, IFFTPrecomputation},
+        polynomial::PolyMultiplier,
+        DensePolynomial,
+        EvaluationDomain,
+        Evaluations as EvaluationsOnDomain,
+    },
     polycommit::sonic_pc::{LabeledPolynomial, PolynomialInfo, PolynomialLabel},
     snark::marlin::{
-        ahp::{verifier, AHPError, AHPForR1CS},
+        ahp::{indexer::CircuitInfo, verifier, AHPError, AHPForR1CS, CircuitId},
+        matrices::MatrixArithmetization,
         prover,
+        witness_label,
         MarlinMode,
     },
 };
+use snarkvm_fields::{batch_inversion_and_mul, PrimeField};
+use snarkvm_utilities::{cfg_iter, cfg_iter_mut, ExecutionPool};
 
 use itertools::Itertools;
 use rand_core::RngCore;
+
+#[cfg(not(feature = "serial"))]
 use rayon::prelude::*;
-use snarkvm_fields::PrimeField;
+
+type Sum<F> = F;
+type Lhs<F> = DensePolynomial<F>;
+type Gpoly<F> = LabeledPolynomial<F>;
 
 impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
-    /// Output the number of oracles sent by the prover in the third round.
-    pub fn num_fourth_round_oracles() -> usize {
-        1
+    /// Output the number of oracles sent by the prover in the fourth round.
+    pub fn num_fourth_round_oracles(circuits: usize) -> usize {
+        circuits * 3
+    }
+
+    /// Output the degree bounds of oracles in the fourth round.
+    pub fn fourth_round_polynomial_info<'a>(
+        circuits: impl Iterator<Item = (CircuitId, &'a CircuitInfo)>,
+    ) -> BTreeMap<PolynomialLabel, PolynomialInfo> {
+        circuits
+            .flat_map(|(circuit_id, circuit_info)| {
+                let non_zero_a_size =
+                    EvaluationDomain::<F>::compute_size_of_domain(circuit_info.num_non_zero_a).unwrap();
+                let non_zero_b_size =
+                    EvaluationDomain::<F>::compute_size_of_domain(circuit_info.num_non_zero_b).unwrap();
+                let non_zero_c_size =
+                    EvaluationDomain::<F>::compute_size_of_domain(circuit_info.num_non_zero_c).unwrap();
+
+                [
+                    PolynomialInfo::new(witness_label(circuit_id, "g_a", 0), Some(non_zero_a_size - 2), None),
+                    PolynomialInfo::new(witness_label(circuit_id, "g_b", 0), Some(non_zero_b_size - 2), None),
+                    PolynomialInfo::new(witness_label(circuit_id, "g_c", 0), Some(non_zero_c_size - 2), None),
+                ]
+                .into_iter()
+                .map(|info| (info.label().into(), info))
+                .collect::<BTreeMap<PolynomialLabel, PolynomialInfo>>()
```

Comment:
since everything gets collected into a `BTreeMap` in the end, are these intermittent collects needed?

Reply:
febb1da7e